Ensure plugin broker works with ephemeral workspaces

amisevsk opened the pull request:
### What does this PR do?
Changes necessary to make plugin broker work correctly when a workspace is marked as ephemeral. This is done by making the broker run as an init container on ephemeral workspaces, so that any volume-related actions are reflected in the workspace's emptyDir volumes.

Detailed changes:
- Slight changes to `EphemeralWorkspaceAdapter`: 1. Make ephemeral workspace check methods static to allow them to be used elsewhere, and 2. add method to "make" a workspace ephemeral, so that all attribute management happens in one place.
- Modify PVC strategies to account for init containers. For `EphemeralWorkspaceAdapter` this is necessary, as otherwise the `/plugins` directory cannot be mounted. I've also included this change for the common and unique strategies, although it doesn't do anything at the moment (since no init containers are added in the non-ephemeral case).
- Apply plugin broker as init container if workspace is ephemeral

### What issues does this PR fix or reference?
#11554 

#### Release Notes
Allow Che 7 workspaces to run as ephermal (emptyDir) containers on Kubernetes/OpenShift.

l0rd commented:
I have a couple of question with respect to the init container approach:
1. Will that work when plugins will run in distinct pods (c.f. https://github.com/eclipse/che/issues/11476 and https://github.com/redhat-developer/rh-che/issues/1005)? I mean ephemeral mode for containers running in distinct pods.
2. Will that work if we want some workspace containers to be loaded first than others (i.e. the editor container is deployed immediately to shorten the start up time for users and other containers are loaded in a lazy manner)?

amisevsk commented:
@l0rd I think ephemeral workspaces will make distinct pods a difficult issue -- an emptyDir volume can only be attached to one pod. It might be complicated but with some architecture work it might be possible. 

![untitled diagram 1](https://user-images.githubusercontent.com/16168279/47825319-9c547f80-dd47-11e8-941d-48f1d7e44d08.png)

As it stands, the diagram above shows what a Che 7 workspace looks like in terms of volumes. From this diagram, to make sure access to projects is shared, theia-ide, che-dev, and any plugins (e.g. che-hello) that needs access to /projects would need to be in the same pod. Further, for the broker to work, it needs to be an init container on the theia pod, so the only plugin that could be separated out into its own pod is the che-machine-exec container.

On the plugin broker side, however, since we only need to touch the plugins volume, the broker container needs only to run as a init container on the theia (or editor) container. If e.g. `/projects` was RWX, then the theia container could be its own pod that starts separately, with the broker as its init container. 

One thing Alex and I talked about briefly is potentially figuring out requirements for multiple brokers (e.g. a theia plugin-only broker would not need an initial run, and could run exclusively as an init container. We didn't really go far down this route though.

garagatyi commented:
@amisevsk FYI we were also considering an ability to create plugins not only for editors but for other items, such as plugins (yeah, plugins of plugins). For example JDT.LS can have its own plugins. But we haven't finished discussions regarding that.
And I would love to remove workspace logs volume in Che 7 since it doesn't make much sense now, here is the [issue](https://github.com/eclipse/che/issues/11806).

garagatyi commented:
@l0rd after discussing the brokering process with Angel and Serhii I'm thinking that we can do improvements described below to tackle your points.
We can set whether a broker changes workspace configuration or copies files. If it just changes configuration we don't need to add volumes to it (actually it can specify that it doesn't need any volume) and it can play with ephemeral workspaces nicely. If it just copies files it would go to the init containers and would probably be able to copy plugins without the usage of non-ephemeral volumes. And only in a case, it needs to do both steps it would be running 2 times in a case of the ephemeral workspace.
Maybe someone would also come up with other improvements for this flow.

l0rd commented:
@garagatyi @amisevsk @skabashnyuk I will try to provide a better description of the requirements. And I think we should have a call today or tomorrow to discuss it.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java:
```diff
@@ -108,6 +110,9 @@ public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
     k8sEnv.getPersistentVolumeClaims().put(pvcName, pvc);
     for (Pod pod : k8sEnv.getPods().values()) {
       PodSpec podSpec = pod.getSpec();
+      List<Container> containers = new ArrayList<>();
```
Good catch -- thank you. I meant to loop over this list.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/EphemeralWorkspaceAdapter.java:
```diff
@@ -61,11 +63,21 @@ public boolean isEphemeral(Workspace workspace) {
    *     is removed for any reason, the data in the `emptyDir` volume is deleted forever
    */
-  public boolean isEphemeral(Map<String, String> workspaceAttributes) {
+  public static boolean isEphemeral(Map<String, String> workspaceAttributes) {
```
Done.

amisevsk commented:
I've addressed the review comments and added tests. I also had to rebase the PR to fix a merge conflict.

amisevsk commented:
ci-test

riuvshin commented:
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1139/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1139//Selenium_20tests_20report/)
docker image: eclipseche/che-server:11786
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

garagatyi commented:
@eclipse/eclipse-che-qa Please, take a look at the report

artaleks9 commented:
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1139/) doesn't show any regression against this Pull Request.

amisevsk commented:
Fixed codacy's complaining and rebased onto current master.

@garagatyi @ibuziuk @sleshchenko I'd like to merge tomorrow unless there's something else that needs to be improved.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesBrokerInitContainerApplier.java:
```diff
@@ -0,0 +1,100 @@
+   * @param workspaceEnvironment
+   * @param runtimeID
+   * @param pluginsMeta
+   * @throws InfrastructureException
```
I found it a bit useless to have listed parameters, exceptions without any description. WDYT?

amisevsk replied:
Good point; I'll fix it.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesBrokerInitContainerApplier.java:
```diff
@@ -0,0 +1,100 @@
+   * @throws InfrastructureException
+   */
+  public void apply(
+      InternalEnvironment workspaceEnvironment,
```
I think that we haven't used `workspaceEnvironment` word anywhere. Please consider changing it to `internalEnvironment`

amisevsk replied:
My intention in using `workspaceEnvironment` was to avoid confusion with `brokerEnvironment` and to support the intuition that we're basically copying environment info from the broker to the workspace. I do agree that immediately casting `workspaceEnvironment` to `kubernetesEnvironment` breaks that idea completely though. 

I've set the type of `workspaceEnvironment` to `E` and removed the cast. WDYT?

sleshchenko replied:
> My intention in using workspaceEnvironment was to avoid confusion with brokerEnvironment

Makes sense

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesBrokerInitContainerApplier.java:
```diff
@@ -0,0 +1,100 @@
+      Collection<PluginMeta> pluginsMeta)
+      throws InfrastructureException {
+
+    KubernetesEnvironment kubernetesEnvironment = (KubernetesEnvironment) workspaceEnvironment;
```
Since your class is generified with `E extends KubernetesEnvironment` - you can avoid casting by putting E as type of `workspaceEnvironment`.
BTW it is not safe to cast without any check of type.

amisevsk replied:
Done.

amisevsk commented:
@sleshchenko Thanks for your review. I agree that running plugin broker twice is not ideal, but also don't know if it will be possible to avoid (at least, for the plugins that require e.g. services on the workspace Pod. One way the init container may be more convenient is for plugins (e.g. theia) that don't require modifying the OpenShift object, where the init container could be sufficient.
